### ISSUE TITLE
acquireTokenSilent returns null for some calls when multiple threads try to access token

### ISSUE THREAD

**Issue by fatihsokmen** [AzureAD/microsoft-authentication-library-for-android#995]
**UPDATE**: Title seems to be misleading. It should read as acquireTokenSilent throws MsalClientException with NO_CURRENT_ACCOUNT message which ends up with null token.

**Describe the bug**
I have an android client api which is called by image loading library to load many photos onto recyclerview viewholder. Bearer token needs be added to api header to request photo file. While many threads try to access token in parallel, some of them get null. 

I did some investigation in sdk source code and realised that following function [persistCurrentAccount] could be the reason why it gets null. It is quite possible that while some threads clear out _sharedPreferencesFileManager_ , and the others can read from it. So it ends up with getting null. 

I've synchronised my function call to fix this issue at the moment but I believe this should be done by sdk and access to **acquireTokenSilent** or **persistCurrentAccount** should be limited one thread at a time.

<img width="839" alt="Screenshot 2020-04-06 at 15 24 08" src="https://user-images.githubusercontent.com/2427299/78571053-ad979a80-781d-11ea-8f20-1f52675c6860.png">

**Smartphone (please complete the following information):**
 - Device: Android Emulator
 - Android Version: 28
 - Browser: Chrome
 - MSAL Version : 1.0.0 to 1.3.0

**Stacktrace**
N/A

**To Reproduce**
1 - Create a recycler view with photo on each row.
2 - Scroll recyclerview.
3 - Photo download should use access token to retrieve data.  

**Expected behavior**
If user logged in, it should always return a non-null token or an appropriate exception.

**Actual Behavior**
_acquireTokenSilent_ sometime returns null.

**Screenshots**
NA

**Additional context**
NA

**Comment by iambmelt** [comment]
@rpdome I've assigned this to you; IIRC you were making some tweaks around this current account query/persistence. Let me know if you need to load balance this

**Comment by iambmelt** [comment]
Thanks for your report @fatihsokmen - we'll look into this threading issue

For your specific scenario: are you accessing the same endpont/resource to load these images? If yes, you can reuse your `Bearer` token across calls up to the lifetime of the token (usually 1 hour)

**Comment by fatihsokmen** [comment]
Hi @iambmelt, thanks for getting back.

Yes, it is the same resource across calls. As a quick fix now, I've synchronised access to resource which works fine now. 

In regards to your advise, managing lifetime in the code can be painful and possibly can raise more issues. I think sdk itself ideally should be single source of truth. Thanks

**Comment by fatihsokmen** [comment]
Hi,
Looks like synchronising access to token has not solved my issue. Concurrent access to token returns null for some threads again.

**Comment by rpdome** [comment]
Hi @fatihsokmen,

Are you getting ILocalAuthenticationResult == 'null' as a result of the acquireTokenSilent operation?

I was debugging the code and noticed that we do have a check for 'null' account in the silent flow (i.e. as the result of persistedAccount due to race condition) - MSAL would throw an exception.

(I'll still fix the synchronization issue, but I just want to make sure that this is not the only issue that is blocking you)

Thanks,
Dome

**Comment by fatihsokmen** [comment]
Apologies. Title seems to be misleading. I checked out the error again and I confirm I am getting **MsalClientException(MsalClientException.NO_CURRENT_ACCOUNT)**, indeed this is not true. 

To be more clear, this is how I produce the error: 
**First Attempt**: Calling **acquireTokenSilent** in parallel (ex: downloading many photos with token in header) threw MsalClientException with NO_CURRENT_ACCOUNT message.

**Second Attempt** - I synchronised access to resource but still didn't seem to work for most cases. 

**My Conclusion** Multiple access to token in a short period of time (ex: in a second) causes MsalClientException(MsalClientException.NO_CURRENT_ACCOUNT). Synchronised access doesn't help as it is supposed to be.

**How I do it now:**
I know that concurrent access to resource isn't helpful which may end up with MsalException. Therefore, I cache the token after each successful read, and next time if it fails, I use cached value which is probably cached in the same second or so. I am still testing this solution and seems to work ok for now, but I need more test to confirm.

```
val accessToken = synchronized(this) {
    val actualToken = runCatching { 
         msal.acquireTokenSilent(SCOPES, AUTHORITY).accessToken 
    }.getOrNull()
    
    if (actualToken != null) {
        cachedToken = actualToken
        actualToken
    } else {
        cachedToken
    }
}
```
